ui-v2/app/components/healthcheck-info.js

```diff
@@ -0,0 +1,14 @@
+import Component from '@ember/component';
+import { subscribe } from 'consul-ui/utils/computed/purify';
+const count = function(value) {
+  if (Array.isArray(value)) {
+    return value.length;
+  }
+  return value;
+};
+export default Component.extend({
+  tagName: '',
+  passingCount: subscribe('passing', count),
+  warningCount: subscribe('warning', count),
+  criticalCount: subscribe('critical', count),
+});
```

Comment:
Unrelated to naming, how do you feel about partial application for the `purify` helper? Something like...

```js
const count = subscribe(function(value) {
  if (Array.isArray(value)) {
    return value.length;
  }
  return value;
});
// ...
passingCount: count('passing')
```

Reply:
Can't these (at least in this case) use the built-in computed macro `alias` ? `passingCount: alias('passing.length')`

Reply:
Alias won't account for the case where `passing` isn't a length but is instead the count value already.

Reply:
aha, thanks - thought there might be a regular input

Reply:
@DingoEatingFuzz , I would like to replace _all_ `computed` with `subscribe` throughout the app. I prefer the pure approach, pass arguments in - return something out, touch as little as possible apart from what is passed in via arguments:

```javascript
computed(
  'passing',
  function() { // missed opportunity to use arguments
    // this is my particular gripe, `subscribe` gives it to me as an argument
    // so I don't even need to import `get`
    return get(this, 'passing')
  }
)
```

If you get time take another read of the PR for the `purify` thing, there's a lot of info there on other reasons why I prefer it.

Anyway time for me to finish up, lets chat Monday, have a good weekend!

Reply:
I have read the purify PR and I honestly don't get the motivation. I see the difference, and I agree with both yours and Tom's reasoning that dependent keys are nothing more than "hints" for Ember, but if this is a problem that will be addressed in Ember core, why take it on as your own problem?

Now what you have is an incomplete leaky abstraction over the top of `Ember.computed` that has no documentation. 

Incomplete because as you mention, the `args.map` tries its hardest to convert dependent keys into the correct object properties, but it doesn't support the entire DSL (not to mention the DSL has ambiguity since it isn't necessarily developer intent).

Leaky because as a developer you still have to be conscious of `Ember.computed`, how it works and the dependent key DSL.

It just doesn't seem pragmatic to attempt to gloss over a cornerstone feature of the framework.

Reply:
Hey @DingoEatingFuzz 

Sorry, I just meant to give it _another_ pass if you had time. The main reason for `purify` is to make the function you pass to `computed` pure (which is why it's called `purify`). From your comments above I didn't get the impression that you'd taken that away so thought it might be useful to take another pass at it. For example:

> Which eliminates the concern of `computed` vs. `subscribed`.

If the main reason/concern for `purify` is to make the function you pass to `computed` pure, then let's use a good definition of a pure function to begin:

> A pure function is a function where the return value is only determined by its input values, without observable side effects

In your code above:

```javascript
const value = this.get(prop);
```

This/`this` isn't an input value. To address your point directly, the concern of `computed` vs `subscribe` isn't eliminated here.

Maybe a familiar example might help:

https://github.com/hashicorp/nomad/blob/f29c3e81a2ab69474c105a85cd8ef4d87830b8f8/ui/app/controllers/jobs/index.js#L135-L177

```javascript
filteredJobs: computed(
    'visibleJobs.[]',
    'selectionType',
    'selectionStatus',
    'selectionDatacenter',
    'selectionPrefix',
    function() {
      const {
        selectionType: types,
        selectionStatus: statuses,
        selectionDatacenter: datacenters,
        selectionPrefix: prefixes,
      } = this.getProperties(
        'selectionType',
        'selectionStatus',
        'selectionDatacenter',
        'selectionPrefix'
      );
      return this.get('visibleJobs').filter(job => {
        ...
      });
    }
  ),
```

could be rewritten as:

```javascript
filterJobs: subscribe(
  'visibleJobs.[]',
  'selectionType',
  'selectionStatus',
  'selectionDatacenter',
  'selectionPrefix',
  function(jobs = [], types = [], statuses = [], datacenters = [], prefixes = []) {
     return jobs.filter(
     	job => {
     	  ...
     	} 
     )
  }
```

My reasons for doing this are that it's much easier to read and follow what's happening, and you get the additional benefit of being able to use default arguments. Good for type hinting and good for avoiding more complicated defensive coding. I can also hoist the function up and out if I want without having to worry about depending on `this` or `get`. I can then reuse it and/or test it in isolation.

Considering we use `computed` a lot in our codebases, which version of the above would you prefer to write multiple times?

I've read over your other points here a few times and wondered how to respond. I think it's important and fair that I touch on your points here, but I don't want to get too far into the weeds. More than happy to chat about this later if you need more clarification.

Overall I think we can all agree with Tom that this is a problematic. I don't see `purify` as a 'solution to a problem' more of a slight improvement.
I've decided to write these ~10-20 lines of code to help me reduce boilerplate, which in turn helps our users. As and when the ember core team release a different approach, I'll consider using that instead:

> We then identify existing tools that simplify the workflow. If a sufficient tool does not exist, we step in to build it. This leads to a fundamentally technology-agnostic view — we will use the best technology available to solve the problem. As technologies evolve and better tooling emerges, the ideal workflow is just updated to leverage those technologies. Technologies change, end goals stay the same.

Your other points:

> - Now what you have is an incomplete leaky abstraction over the top of `Ember.computed` that has no documentation.
> - Incomplete because as you mention, the `args.map` tries its hardest to convert dependent keys into the correct object properties
> - Leaky because as a developer you still have to be conscious of `Ember.computed`, how it works and the dependent key DSL.
> - It just doesn't seem pragmatic to attempt to gloss over a cornerstone feature of the framework.

In my eyes these are [Straw Man](https://en.wikipedia.org/wiki/Straw_man) arguments.

1. Incomplete

```javascript
// Right now this just takes the first part of the path so:
// `items.[]` or `items.@each.prop` etc
// gives you `items` which is 'probably' what you expect
// it won't work with something like `item.objects.[]`
// it could potentially be made to do so, but we don't need that right now at least
```

_it could potentially be made to do so, but we don't need that right now at least_

We don't need it so we didn't add it. This is not something that is meant to be shared, we are the only authors and consumers. It is 'complete' for what we want to use it for. Have we finished working on this? Maybe, maybe not. The only thing it doesn't support is Ember's unique syntax. If we want to use that we could just use `alias` first.

As a comparison, as far as I'm aware Ember angle bracket syntax isn't supported during testing, and doesn't support all types of components for example `{{link-to}}`. IMO this is more 'incomplete' - it doesn't stop anyone from using it though.

So, it's not incomplete, I could add extra features that I don't need, yes, but not 'incomplete'.

2. Leaky 😱 

This one really confused me.

> Leaky because as a developer you still have to be conscious of `Ember.computed`, how it works and the dependent key DSL.

...because as a developer I have to be conscious of `Ember.computed`? I don't understand completely how this phrase defines the term 'leaky'? Leaky as in abstraction leaky? As in leaking the details of the underlying implementation? 

```javascript
// passingCount subscribes to changes to `passing` and counts the new value
passingCount: subscribe('passing', count)
```
**passingCount** **subscribes** to changes to **passing** and **count**s the value

I don't see how the underlying implementation is 'leaked' anywhere here. Am I missing something? If I change the name of `count` to `lengthOrValue` instead, then yes that would be leaking the implementation of counting into the abstraction (in this case, the name of the function).

I am aware that it happens to use `computed` just because I wrote it, but I guess I could implement what I've done using something else, like traditional EventListeners or maybe ember observables. I wouldn't have to change how I would use this if I did so.

There is no leaky abstraction here. There is no mention of `computed`, `get` or actually anything to do with Ember. It could be using `computed`, or observables, or maybe even `EventSource`s - and you'd never know.

Maybe I've misunderstood what you meant.

3. Tries its hardest 🙀 

> the `args.map` tries its hardest to convert dependent keys into the correct object properties, but it doesn't support the entire DSL

I'm not sure what you meant by 'tries its hardest'? Sounds terrible! [One of the aims of a Straw Man argument is to cause alarm](https://en.wikipedia.org/wiki/Straw_man#Examples). The code here doesn't 'try its hardest', it doesn't try at all, hopefully the comment in the code explains that.

4. Has no documentation

Most of the Consul UI is undocumented, the same for Vault and Nomad. For example, I've also noticed you do something very similar here without any documentation:

https://github.com/hashicorp/nomad/blob/f29c3e81a2ab69474c105a85cd8ef4d87830b8f8/ui/app/controllers/clients/index.js#L40-L43
https://github.com/hashicorp/nomad/blob/f29c3e81a2ab69474c105a85cd8ef4d87830b8f8/ui/app/utils/qp-serialize.js

To be fair, here you are factually correct that there is no real documentation for this function. I'd be happy to add some if you think its useful. But I think that in the scheme of things this just isn't really important, which is why I'd still see this as "attacking a [Straw Man]( https://en.wikipedia.org/wiki/Straw_man)"

Lastly, there is one and only one reason why I am hesitant to use `purify` and that is because at some point we'll be able to write code like:

```javascript
export default class TodoList extends Component {
  @computed('todos.length')
  get uncompletedTodos() {
    return this.todos.filter(todo => !todo.completed);
  }
}
```

My concern is the upgrade path from `subscribe` to be able to do this. I've considered this and decided that I am still ok using `subscribe`. Why?:

1. If I don't tie any of my code to `subscribe` say by wrapping it up like you have done above, instead I compose functionality. I can drop back easily if need be:

```javascript
export default class TodoList extends Component {
  @computed('passing')
  get passingCount () { // you can't pass arguments to a getter
    return count(this.passing);
  }
}
```

2. Potentially I could write a `@subscribe` decorator, like other people have written their onw decorators.
3. I've no idea when decorators are going to land _and_ I can use them. In my experience it can be at least a year of waiting (I'm not complaining). I get the feeling that this _might_ be quicker, but who knows 🤷‍♂️ 

Hopefully that helps to explain where I'm coming from. Feel free to ping me offline if you'd like to discuss further.

Thanks

John

Reply:
Skipping over your explanation of what purify is, since I already told you I know what it is, what it does, and how it works, I'll restate my case on the leaky and incomplete parts.

As I already mentioned, it's leaky because purify + computed (aka subscribe) still relies on the user of subscribe to understand ember's DK syntax and semantics. Thus the details of the underlying behavior (computed) leaks.

I also already stated my case about the abstraction being incomplete, but I'll try to be clearer this time. When I say "tries its best" I am mostly referring to this comment directly from the purify source.

> it won't work with something like `item.objects.[]`

Sure, you don't need this so you didn't make it work, but the fact remains that to no fault of your own, dependent keys can't be perfectly mapped to the properties the author of a computed property may want to use.

Which brings us to the actual point of my comment that you neglected to mention in your response: I don't understand the motivation and it isn't pragmatic.

Computed properties are member functions, so why bother making them pure in the first place? Computed property macros (which are in fact [a documented feature of ember](https://guides.emberjs.com/v2.13.0/object-model/computed-properties/#toc_computed-property-macros)) generate member functions. This is the intended way to DRY out common computed properties (as further evidenced by the framework provided macros such as `alias`, `collect`, etc.).

And since computed properties are already a thing, and computed property macros are already a thing, and the remaining blemish of having to declare dependent keys is already a problem being addressed by the framework, there is no case to be made for introducing your own replacement being pragmatic.